src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1013 @@
+
+    // Stack maintaining whether or not the node we're serializing inherits from the underlying BSON
+    // data base class required for classes that use bsoncxx view types.
+    std::stack<bool> _curNodeInheritsUnderlyingBSONDataBase;
+
+    // Boolean value that tracks whether or not the current root BSON document or array is being
+    // treated as a root element.
+    bool _objAsRootElement;
+
+};  // BSONOutputArchive
+
+class BSONInputArchive : public InputArchive<BSONInputArchive> {
+    enum class InputNodeType { InObject, InEmbeddedObject, InEmbeddedArray, InRootElement };
+
+   public:
+    /**
+    * Construct a BSONInputArchive from an input stream of BSON data.
+    *
+    * @param stream
+    *    The stream from which to read BSON data.
+    */
+    BSONInputArchive(std::istream& stream)
+        : InputArchive<BSONInputArchive>(this),
+          _nextName(nullptr),
+          _readStream(stream),
+          _readFirstDoc(false) {
+    }
+
+   private:
+    /**
+     * Reads the next BSON document from the istream. This should be called whenever
+     * we are starting to load in a root element or root node.
+     */
+    void readNextDoc() {
+        // Determine the size of the BSON document in bytes.
+        // TODO: Only works on little endian.
+        int32_t docsize;
+        char docsize_buf[sizeof(docsize)];
+        _readStream.read(docsize_buf, sizeof(docsize));
+        std::memcpy(&docsize, docsize_buf, sizeof(docsize));
+
+        // Throw an exception if the end of the stream is prematurely reached.
+        if (_readStream.eof() || !_readStream || docsize < 5) {
+            throw cereal::Exception("No more data in BSONInputArchive stream.\n");
+        }
+
+        // Create a shared buffer to store the BSON document.
+        _curBsonData =
+            std::shared_ptr<uint8_t>{new uint8_t[docsize], [](uint8_t* p) { delete[] p; }};
+
+        // Read the BSON data from the stream into the buffer.
+        std::memcpy(_curBsonData.get(), docsize_buf, sizeof(docsize));
+        _readStream.read(reinterpret_cast<char*>(_curBsonData.get() + sizeof(docsize)),
+                         docsize - sizeof(docsize));
+
+        // Make sure there were no errors reading the BSON data.
+        if (_readStream.eof() || !_readStream) {
+            throw cereal::Exception("No more data in BSONInputArchive stream.\n");
+        }
+
+        // Store the BSON data of the document in a view that we can access.
+        _curBsonDoc = bsoncxx::document::view{_curBsonData.get(), static_cast<size_t>(docsize)};
+        _curBsonDataSize = docsize;
+
+        // Specify that we've read a document.
+        _readFirstDoc = true;
+    }
+
+    /**
+     * Searches for the next BSON element to be retrieved and loaded.
+     *
+     * @return the element with _nextName as a key, or if the current node is an
+     *         array, the next element in that array.
+     */
+    inline bsoncxx::types::value search() {
```

Comment:
It helps a little bit, but doesn't deal with the fact that we still need to page in the values behind the keys (so we'd save on scanning, but we'd still thrash the cache).

I guess the idea that I have in mind is that as you traverse the fields in the struct (that cereal knows about), you push the requests into a vector.  When you go into a sub doc, you push a "sub doc token".  Then when you leave a doc, you go back in the list of tokens, looping over the bson once, and fill out the instructions by linearly scanning names until you find the open doc token.

I.e.

```cpp
/* Given input data:
{
  foo : 1
  bar : {
    baz : 2
  },
  qux : 3
};
*/

struct Bar {
  int32_t baz;
  template <class Archive>
  void serialize(Archive& ar) {
    ar(CEREAL_NVP(baz)); 
  };
};

struct Data {
  int32_t foo;
  Bar bar;
  int32_t qux;

  template <class Archive>
  void serialize(Archive& ar) {
    ar(CEREAL_NVP(foo), CEREAL_NVP(bar), CEREAL(qux)); 
  };
};

/* I'd expect:
push("foo", types_enum::int32, &foo);
push("bar", types_enum::enter_object);
push("baz", types_enum::int32, &baz);
push(exit_object);
push("qux", types_enum:int32, &qux);

Then when we loop over the bson object, you scan all of the requirements at the correct level (you skip subobject fields) and fill them if you match.  When you hit an enter object, you scan only the records at the correct level until your corresponding exit_object.

As you find records, you erase them from the array.

I suspect this will be faster than a set for numKeys < 1000.
*/
```